src/chainstate/burn/operations/peg_out_fulfill.rs

```diff
@@ -0,0 +1,228 @@
+use stacks_common::codec::StacksMessageCodec;
+use stacks_common::types::chainstate::StacksBlockId;
+
+use crate::burnchains::BurnchainBlockHeader;
+use crate::burnchains::BurnchainTransaction;
+use crate::chainstate::burn::Opcodes;
+use crate::types::chainstate::StacksAddress;
+use crate::types::Address;
+
+use crate::chainstate::burn::operations::Error as OpError;
+use crate::chainstate::burn::operations::PegOutFulfillOp;
+
+impl PegOutFulfillOp {
+    pub fn from_tx(
+        block_header: &BurnchainBlockHeader,
+        tx: &BurnchainTransaction,
+    ) -> Result<Self, OpError> {
+        if tx.opcode() != Opcodes::PegOutFulfill as u8 {
+            warn!("Invalid tx: invalid opcode {}", tx.opcode());
+            return Err(OpError::InvalidInput);
+        }
+
+        let (amount, recipient) = if let Some(Some(recipient)) = tx.get_recipients().first() {
+            (recipient.amount, recipient.address.clone())
+        } else {
+            warn!("Invalid tx: Output 2 not provided");
+            return Err(OpError::InvalidInput);
+        };
+
+        let chain_tip = Self::parse_data(&tx.data())?;
+
+        let txid = tx.txid();
+        let vtxindex = tx.vtxindex();
+        let block_height = block_header.block_height;
+        let burn_header_hash = block_header.block_hash;
+
+        Ok(Self {
+            chain_tip,
+            amount,
+            recipient,
+            txid,
+            vtxindex,
+            block_height,
+            burn_header_hash,
+        })
+    }
+
+    fn parse_data(data: &[u8]) -> Result<StacksBlockId, ParseError> {
+        /*
+            Wire format:
+
+            0      2  3                     35
+            |------|--|---------------------|
+             magic  op       Chain tip
+
+             Note that `data` is missing the first 3 bytes -- the magic and op must
+             be stripped before this method is called. At the time of writing,
+             this is done in `burnchains::bitcoin::blocks::BitcoinBlockParser::parse_data`.
+        */
+
+        if data.len() < 32 {
+            warn!(
+                "PegInOp payload is malformed ({} bytes, expected at least {})",
+                data.len(),
+                32
+            );
+            return Err(ParseError::MalformedData);
+        }
+
+        StacksBlockId::from_bytes(data).ok_or(ParseError::MalformedData)
```

Comment:
You should take the 32 byte slice from `data` here.  This operation should be valid even if the OP_RETURN contains data after the 35th byte.  We'd reserve bytes 35-80 for future use.

Reply:
Done :heavy_check_mark: